src/inmanta/agent/executor.py

```diff
@@ -289,27 +404,137 @@ def initialize_envs_directory() -> str:
+                LOGGER.debug("%s's cleanup job is already cancelled!", self.__class__.__name__)
+            finally:
+                self.cleanup_job = None
+
+    @abc.abstractmethod
+    async def get_pool_members(self) -> Sequence[PoolMember]:
+        """
+        Retrieve the members of the pool: those that need to be checked for the cleanup.
+        """
+        pass
+
+    def clean_pool_member_from_manager(self, pool_member: PoolMember) -> None:
+        """
+        Additional cleanup operation(s) that need to be performed by the Manager regarding the pool member being cleaned.
+        This method assumes to be in a lock to prevent other operations to overlap with the cleanup.
+        """
+        pass
+
+    async def cleanup_inactive_pool_members(self) -> float:
+        """
+        Cleans up idle pool member
+
+        :return: When to run the cleaning next time: default retention time or lowest expiration time of one of the pool members
+        """
+        loop = asyncio.get_running_loop()
+
+        cleanup_start = datetime.datetime.now().astimezone()
+        # Number of seconds relative to cleanup_start
+        run_next_cleanup_job_at: float = float(self.retention_time)
+        pool_members = await self.get_pool_members()
+        for pool_member in pool_members:
+            try:
+                if pool_member.can_be_cleaned_up(self.retention_time):
+                    async with self._locks.get(pool_member.get_id()):
+                        # Check that the executor can still be cleaned up by the time we have acquired the lock
+                        if pool_member.can_be_cleaned_up(self.retention_time):
+                            await loop.run_in_executor(self.thread_pool, pool_member.clean)
+                            self.clean_pool_member_from_manager(pool_member)
+                else:
+                    # If this pool member expires sooner than the cleanup interval, schedule the next cleanup on that timestamp.
+                    run_next_cleanup_job_at = min(
+                        run_next_cleanup_job_at,
+                        (
+                            datetime.timedelta(seconds=self.retention_time) - (cleanup_start - pool_member.last_used())
+                        ).total_seconds(),
+                    )
+            except Exception:
+                LOGGER.exception(
+                    "An error occurred while cleaning the %s pool member `%s`",
+                    self.__class__.__name__,
+                    pool_member.get_id(),
+                )
+
+        cleanup_end = datetime.datetime.now().astimezone()
+
+        return max(0.0, run_next_cleanup_job_at - (cleanup_end - cleanup_start).total_seconds())
+
+    async def cleanup_inactive_pool_members_task(self) -> None:
+        """
+        This task periodically cleans up idle pool member
+        """
+        while self.running:
+            await asyncio.sleep(await self.cleanup_inactive_pool_members())
```

Comment:
I would find it easier to read if the `cleanup_inactive_pool_members` was part of this method as well (now the sleep is an odd place)